docs/smart_contracts.md

```diff
@@ -0,0 +1,183 @@
+- [Lucet](https://github.com/bytecodealliance/lucet/), the Sandboxing WebAssembly Compiler
+
+ABI Bindings (with our Massa crate)
+
+- [Experimental WASI API bindings for Rust](https://crates.io/crates/wasi)
+
+### Specs
+
+#### WASM
+
+- https://www.w3.org/TR/wasm-core/
+- https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#type-section
+
+#### WASI
+
+https://wasi.dev/
+> Note that everything here is a prototype, and while a lot of stuff works, there are numerous missing features and some rough edges. For example, networking support is incomplete.
+
+#### WASM Interfaces Types
+
+https://github.com/WebAssembly/interface-types/blob/main/proposals/interface-types/Explainer.md
+> Note: This repository is now archived. The interface types proposal has changed quite a lot since this was first implemented and the trajectory of the proposal and the implementation within wasmtime no longer aligns with this repository.
+
+https://hacks.mozilla.org/2019/08/webassembly-interface-types/
+
+#### WITX
+
+https://github.com/WebAssembly/WASI/blob/main/docs/witx.md
+
+### Wasmer
+
+https://wasmer.io/
+
+- WASI: https://docs.rs/wasmer-wasi/ & https://github.com/wasmerio/wasmer/blob/master/examples/wasi.rs
+- Interfaces Types: https://github.com/wasmerio/interface-types
+
+```rust
+use wasmer::{imports, Function, Instance, Module, Store, Value};
+
+type Address = u64;
+type Amount = u64;
+
+fn main() -> anyhow::Result<()> {
+    let module_wat = r#"
+    ;; TODO: put some Python interpreter here!
+    (module
+    (type $t0 (func (param i32) (result i32)))
+    (func $add_one (export "add_one") (type $t0) (param $p0 i32) (result i32)
+        get_local $p0
+        i32.const 1
+        i32.add))
+    "#;
+
+    let store = Store::default();
+    let module = Module::new(&store, &module_wat)?;
+
+    // The module could import any massa specific API calls
+    let import_object = imports! {
+        "env" => {
+            "massa_get_balance" => Function::new_native(&store, massa_get_balance)
+        },
+    };
+
+    fn massa_get_balance(_: Address) -> Amount {
+        todo!()
+    }
+
+    let instance = Instance::new(&module, &import_object)?;
+    // this will execute the smart-contract "add-one"
+    let add_one = instance.exports.get_function("add_one")?;
+    // the context passed here is `&[Value::I32(42)]`
+    let result = add_one.call(&[Value::I32(42)])?;
+    assert_eq!(result[0], Value::I32(43));
+
+    Ok(())
+}
+```
+
+Tools
+
+- https://wapm.io
+- https://webassembly.sh/
+- https://github.com/wasienv/wasienv
+
+## How to pass/receive Rust types to/from wasm?
+
+https://github.com/adrien-zinger/wasm-simulation/issues/2
+
+### Solution A: use AssemblyScript little overhead
+
+- https://github.com/adrien-zinger/wasmer-as
+- https://github.com/near/assemblyscript-json
+
+### Solution B: use WasiEnv
+
+TODO
+
+### Solution C: use WasmBindgen
+
+TODO
```

Comment:
I was mistaking saying Solana SC engine was based somehow on https://github.com/rustwasm/wasm-pack

It smartly uses https://fr.wikipedia.org/wiki/BSD_Packet_Filter as shown in this example https://github.com/mvines/solana-bpf-program-template